doc/man3/EVP_SKEY_new.pod

```diff
@@ -0,0 +1,69 @@
+=pod
+
+=head1 NAME
+
+EVP_SKEY,
+EVP_SKEY_new,
+EVP_SKEY_new_raw_key,
+EVP_SKEY_up_ref,
+EVP_SKEY_free
+- opaque symmetric key allocation and handling functions
+
+=head1 SYNOPSIS
+
+ #include <openssl/evp.h>
+
+ typedef evp_skey_st EVP_SKEY;
+
+ EVP_SKEY *EVP_SKEY_new(OSSL_LIB_CTX *libctx, const char *keymgmtname, const char *propquery);
+ EVP_SKEY *EVP_SKEY_new_raw_key(const unsigned char *key, size_t keylen);
+ int EVP_SKEY_up_ref(EVP_SKEY *key);
+ void EVP_SKEY_free(EVP_SKEY *key);
+
+=head1 DESCRIPTION
+
+B<EVP_SKEY> is a generic structure to hold symmetric keys as opaque objects.
+The keys themselves are often referred to as the "internal key", and are handled by
+providers through L<EVP_KEYMGMT(3)>.
+
+Conceptually, an B<EVP_SKEY> internal key may hold a symmetric key, and along
+with those, key parameters if the key type requires them.
+
+The EVP_SKEY_new() function allocates an empty B<EVP_SKEY> structure which is
```

Comment:
@davidben So you propose to combine EVP_SKEY_new() with EVP_SKEY_import(), right?
I'm not sure I see how we can make one allocation this way but yes, it can be done

Reply:
Yeah. More generally, any operation that would fill in an empty `EVP_SKEY` can return a new `EVP_SKEY`.

As for saving an allocation, it'll depend a bit on how the many layers underneath in the providers mess work. But even if providers today can't support that, you all at least will have the option to extend the providers API later to do so. But broadly, you can imagine making one allocation for M+N bytes where M is the size of the `EVP_SKEY` fields and then N is the size of the provider/key-specific data.

So for something like `EVP_MD_CTX`, it would be as if you allocated a:

```
struct EVP_MD_CTX_SHA256 {
  EVP_MD_CTX base;
  SHA256_CTX sha256;
};
```

And then cast that to `EVP_MD_CTX*`. This is how most languages implement inheritance. Of course, we cannot do this optimization in `EVP_MD_CTX` because OpenSSL design that API with an empty state, but no sense in repeating this. This only works if the type of an individual `EVP_SKEY` cannot change after it is created.

Even without this optimization, this design removes the empty state from the type system. That means that functions (inside and outside OpenSSL) which take `EVP_SKEY` don't need to add extra checks for the case that the `EVP_SKEY` doesn't exist. You know that the type exists, method tables are non-null, etc.

Reply:
wholeheartedly agree, I also would like to avoid empty status as it makes it awkward for the pkcs11-provider to create empty objects that needs to be checked and re-checked to see if they are empty or hold actual content ... the less APIs like that we have the better.

Reply:
@beldmit's approach is consistent with the other like structures in OpenSSL and in general we aim to remain consistent with other parts of the API (while recognising there is already enough inconsistency that you can find an API pattern to support pretty much any approach). 

You'll note that looking back at the referenced items going through the APIs and evolving them in a different direction is a post-3.x item (i.e. 4.0).

Reply:
> wholeheartedly agree, I also would like to avoid empty status as it makes it awkward for the pkcs11-provider to create empty objects that needs to be checked and re-checked to see if they are empty or hold actual content ... the less APIs like that we have the better.

Considering that the complaint is attached to `EVP_SKEY_new()`, I've assumed that the "empty state" is when the resulting `EVP_SKEY` instance doesn't have an actual key assigned.  Looking at the code for `EVP_SKEY_new()`, I can't see that the provider implementation is invoked at all, it's only fetched.

Is this about how `EVP_SKEY_import()` works, i.e. this construct:

``` C
    if ((keydata = evp_keymgmt_newdata(skey->keymgmt)) == NULL
        || !evp_keymgmt_import(skey->keymgmt, keydata, OSSL_KEYMGMT_SELECT_ALL, params)) {
```

Reply:
@levitte the critical issue is calling evp_keymgmt_newdata() first, which requires the provider to return an immutable but valid pointer, ie it requires the provider to create an internal representation of an empty key, which in turns requires a ton of case handling in all other provider functions, to ensure we have something in the key when later invoked.

And it it unclear why you should have these calls back-to-back (evp_keymgmt_newdata + evp_keymgmt_import) which could be expensive on the provider side, except for "tradition", when a single call would do just as well.

Providers should not allow you to import other data into an existing key anyway so the import() function also need to validate that an empty key was provided and error out otherwise [and yet the ECDH code (with PKEYs) expects just that (to import a public point over a half formed key) and it is already a major problem as the amount of checks that I had to embed in the provider after I figured out what the heck was going on is ugly and excessive].

So, essentially, import() can only ever give you a "new" key and all this code is redundant and error prone for no good reason, it just invites bad practices.

Reply:
Right, these are two instances of the general problem with this empty-state pattern that OpenSSL likes to use. By adding these two steps to the interface, both sides of the interface boundary gain complexity for no good reason:

* On the caller side, you have to deal with failure conditions of two objects, and remember that OpenSSL objects do not capture invariants in their types
* On the callee side, you have to implement this extra state and keep track of it, as @simo5 mentions. The missed optimization I describe is one example of what you lose, but more importantly it's just extra complexity all around.

Leaving this to 4.x does not seem like it helps OpenSSL users in any way. I think this is clear to see once one thinks about what this topic actually entails: you would add APIs that combine new-empty + init-from-empty like, have folks migrate to the combined ones, and then you remove the split. The first step is a fully compatible addition and can be done incrementally, one API at a time. Delaying the compatible step just means OpenSSL users cannot prepare ahead of time, and causes more work for them down the line.

Really this pattern should have been abandoned when you all switched from stack-allocatable types to opaque types in 1.1.0. When the types were stack-allocated, it made sense to have an empty state because you needed an infallible way for the caller to get an `EVP_MD_CTX` into _some_ valid state for `EVP_MD_CTX_cleanup`. And thus the `EVP_MD_CTX_init` / `EVP_MD_CTX_cleanup` pair and separate `EVP_DigestInit_ex`.

Once the types are heap-allocated, this is moot. The type on the stack is a pointer, and the infallible initialization step is simply `= NULL`. At that point, the empty state was unnecessary, and one could simply have made the `EVP_DigestInit_ex` analog return the `EVP_MD_CTX *`.

(I've noticed this pattern in OpenSSL of bugs I've filed. I usually try to write my suggestions such that they can be done incrementally. That's how I like to work, and I suggest things based on my experience on a different, related project. #5158, #6640, and #1319 come to mind. I think you all have a habit of interpreting them as big atomic things and they get stuck on the automatic bug deferral train, when they're usually quite straightforward.)

Reply:
> Delaying the compatible step just means OpenSSL users cannot prepare ahead of time, and causes more work for them down the line.

I should add: In the case of a new API like `EVP_SKEY`, the deferral proposal is doing _even worse_ by OpenSSL's users! Rather than giving them the API that would be better for them from day one, deferral would first give them a worse API and then force them to migrate later.

If we already know what the right API is, just add the right API from the start.

Reply:
> If we already know what the right API is, just add the right API from the start.

And that is the actual question - we need to decide which API patterns we want to have in place and what sort of impact it will have on various existing APIs and a migration path.  Making different decisions on an PR-by-PR or API-by-API case makes little sense. Either we figure out what pattern we want to move to and start moving there or we leave things consistent - and we need to make a decision on what API tradeoffs we want.

That has been the overall approach taken to date - each time we run into suggestions like this we don't get to a consensus view on the approach going forward so the can gets kicked down the road. It is not that things get ignored - your suggestions always lead to interesting and at times length discussions - but rarely is a consensus viewpoint reached.

We also have the requirement to support all the existing APIs longer term and there are implications in various changes. We had significant feedback at various times when making what seemed to be simple and positive changes to APIs where it broke actual usage and user dependencies on particular things. It leads to relatively cautious decision making on making API approach changes.

The other thing to keep in mind is that the majority of applications don't have the luxury of making a decision to only support the latest OpenSSL version - most applications end up carrying out a whole range of OpenSSL versions supported in the code. When you look at the API changes over the various versions and what they look like in applications you also tend to consider changes which involve having to significantly restructure code supporting multiple OpenSSL versions.

This is just to communicate some of the discussion that happen around such suggestions ... so there is more transparency/visibility on what happens with your helpful feedback - which is very much welcome and appreciated.